### PR TITLE
Unit Test Case for router

### PR DESCRIPTION
**Pull request by uidoyen**
Closes  [RHOAIENG-2164](https://issues.redhat.com/browse/RHOAIENG-2164)

## Description

Unit Test Case for **router.ts**

## How Has This Been Tested?

`npm run test`

## Test Impact

Screenshot 2024-01-25 at 1 03 33 AM

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 

- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

File: frontend/src/utilities/__tests__/router.spec.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    // Simulate an existing query parameter with the same value
+    window.history.pushState({}, '', `?${key}=${value}`);
+    // Call the hook to set the query parameter with the same value
+    testHook(setQueryArgument)(navigate, key, value);
+
+    // Check if navigate is not called
+    expect(navigate).not.toHaveBeenCalled();
+  });
+  describe('removeQueryArgument', () => {
```
Comment (christianvogt) [inline frontend/src/utilities/__tests__/router.spec.ts:37]: Do not nest these two `describe`. Use separate ones per function being tested.
```suggestion
  });
});
describe('removeQueryArgument', () => {
```

File: frontend/src/utilities/__tests__/router.spec.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    const key = 'param';
+    const value = 'example';
+    // Ensure the query parameter is already set to the same value initially
+    const initialUrl = new URL(window.location.href);
+    initialUrl.searchParams.set(key, value);
+
+    // Simulate an existing query parameter with the same value
+    window.history.pushState({}, '', `?${key}=${value}`);
+    // Call the hook to set the query parameter with the same value
+    testHook(setQueryArgument)(navigate, key, value);
```
Comment (christianvogt) [inline frontend/src/utilities/__tests__/router.spec.ts:32]: `setQueryArgument` is a simple function. Do not use `testHook`.

File: frontend/src/utilities/__tests__/router.spec.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+describe('setQueryArgument', () => {
+  it('should set a query argument and navigate', () => {
+    // Define the key and value for the query parameter
+    const key = 'param';
+    const value = 'example';
+    // Ensure the query parameter does not exist initially
+    const initialUrl = new URL(window.location.href);
+    initialUrl.searchParams.set(key, value);
+    // Call the hook to set the query parameter
```
Comment (christianvogt) [inline frontend/src/utilities/__tests__/router.spec.ts:15]: This is NOT a hook.

File: frontend/src/utilities/__tests__/router.spec.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    // Simulate an existing query parameter with the same value
+    window.history.pushState({}, '', `?${key}=${value}`);
+    // Call the hook to set the query parameter with the same value
+    testHook(setQueryArgument)(navigate, key, value);
+
+    // Check if navigate is not called
+    expect(navigate).not.toHaveBeenCalled();
+  });
+  describe('removeQueryArgument', () => {
+    it('should remove a query argument and navigate', () => {
```
Comment (christianvogt) [inline frontend/src/utilities/__tests__/router.spec.ts:38]: This tests relies on the setup of `key` param from the previous test. Tests should be setup to function in isolation.

File: frontend/src/utilities/__tests__/router.spec.ts (hunk 3 above)
Comment (uidoyen) [reply frontend/src/utilities/__tests__/router.spec.ts:15]: Thanks, @christianvogt, for pointing this out. I apologize for this oversight.

File: frontend/src/utilities/__tests__/router.spec.ts (hunk 2 above)
Comment (uidoyen) [reply frontend/src/utilities/__tests__/router.spec.ts:32]: Thanks, @christianvogt, for pointing this out. I apologize for this oversight.

File: frontend/src/utilities/__tests__/router.spec.ts (hunk 1 above)
Comment (christianvogt) [reply frontend/src/utilities/__tests__/router.spec.ts:36]: You haven't made this change.

File: frontend/src/utilities/__tests__/router.spec.ts (hunk 4 above)
Comment (christianvogt) [reply frontend/src/utilities/__tests__/router.spec.ts:37]: This test _still_ relies on the previous test to set the `key` `param=example`. Run this test in isolation to see the failure.

These lines DO NOT update `window.location` it only updates the `initialUrl` variable which is a _copy_.
```
      const initialUrl = new URL(window.location.href);
      initialUrl.searchParams.set(key, 'example');
```

**Comment by uidoyen** [comment]
@christianvogt Updated the PR accordingly. Thanks

**Review by pnaik1 — DISMISSED** [review]

File: frontend/src/utilities/__tests__/router.spec.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    value: originalLocation,
+  });
+});
+
+describe('setQueryArgument', () => {
+  it('should set a query argument and navigate', () => {
+    const key = 'param';
+    const value = 'example';
+
+    Object.defineProperty(globalThis, 'window', {
```
[the remark this answers is not in the document]
Comment (uidoyen) [reply frontend/src/utilities/__tests__/router.spec.ts:18]: @pnaik1 If we remove that line we would essentially not be resetting the window object back to its original state after the test. by keeping the `Object.defineProperty` block, you ensure that the `window` object is restored to its original state after the test.

[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/src/utilities/__tests__/router.spec.ts:18]: It doesn't seems like this is necessary. However a good test should assign a valid value to assert beforehand. It's not ideal to be assigning and unknown value here though.

I think we need better support for mocking window location. Let's put this through and we can improve on this in the future.

I propose we go ahead with the PR as is and update it when we have better mocking support for window location.

**Pull request merged**